src/mincraft/mincraft.cc

```diff
@@ -0,0 +1,1210 @@
+    {
+        auto primitives = FrameSyncPrimitives{
+            .image_availible_semaphore = logical_device.createSemaphoreUnique( {} ),
+            .render_finished_semaphore = logical_device.createSemaphoreUnique( {} ),
+            .in_flight_fence = logical_device.createFenceUnique( { .flags = vk::FenceCreateFlagBits::eSignaled } ) };
+
+        frame_render_info.sync_primitives.push_back( std::move( primitives ) );
+
+        auto buffer = buffer_builder.withSize( sizeof( UniformBufferObject ) )
+                          .withUsage( vk::BufferUsageFlagBits::eUniformBuffer )
+                          .make( manager );
+
+        frame_render_info.uniform_buffers.push_back( std::move( buffer ) );
+    }
+
+    frame_render_info.imgui_command_buffers = command_pool.createCmdBuffers( static_cast<uint32_t>( frames ) );
+
+    return frame_render_info;
+}
+
+vkwrap::Image
+createDepthBuffer( const vkwrap::Swapchain& swapchain, ranges::range auto&& queues, vkwrap::Mman& manager )
+{
+    auto [ width, height ] = swapchain.getExtent();
+    auto builder = vkwrap::ImageBuilder{};
+
+    auto depth_image = builder.withExtent( { .width = width, .height = height, .depth = 1 } )
+                           .withFormat( vk::Format::eD32Sfloat )
+                           .withTiling( vk::ImageTiling::eOptimal )
+                           .withImageType( vk::ImageType::e2D )
+                           .withQueues( queues )
+                           .withSampleCount( vk::SampleCountFlagBits::e1 )
+                           .withArrayLayers( 1 )
+                           .withUsage( vk::ImageUsageFlagBits::eDepthStencilAttachment )
+                           .make( manager );
+
+    return depth_image;
+}
+
+bool
+shouldRecreateSwapchain( vk::Result result )
+{
+    return ( result == vk::Result::eErrorOutOfDateKHR || result == vk::Result::eSuboptimalKHR );
+}
+
+class KtxError : public std::runtime_error
+{
+  public:
+    explicit KtxError( ktx_error_code_e ec )
+        : std::runtime_error{ ktxErrorString( ec ) },
+          m_ec{ ec }
+    {
+    }
+
+    auto getErrorCode() { return m_ec; }
+
+  private:
+    ktx_error_code_e m_ec;
+};
+
+void
+checkKtxResult( ktx_error_code_e ec )
+{
+    if ( ec != KTX_SUCCESS )
+    {
+        throw KtxError{ ec };
+    }
+}
+
+class UniqueKtxTexture
+{
+    static ktxTexture* createHandle( const std::filesystem::path& filepath )
+    {
+        ktxTexture* handle = nullptr;
+        auto res = ktxTexture_CreateFromNamedFile( filepath.c_str(), KTX_TEXTURE_CREATE_LOAD_IMAGE_DATA_BIT, &handle );
+        checkKtxResult( res );
+        return handle;
+    }
+
+  public:
+    explicit UniqueKtxTexture( const std::filesystem::path& filepath )
+        : m_handle{ createHandle( filepath ) }
+    {
+    }
+
+    [[nodiscard]] ktxTexture* get() const { return m_handle.get(); }
+    ktxTexture* operator->() { return get(); }
+
+    ktx_uint8_t* getData() { return ktxTexture_GetData( m_handle.get() ); }
```

Comment:
It seems that `getData` should be `const` method and return `const ktx_uint8_t*`.